alpenhorn/queue.py

```diff
@@ -0,0 +1,426 @@
+            The name of the FIFO to return the size of.
+
+        Returns
+        -------
+        fifo_size : int
+            The size of the FIFO as explained above.
+        """
+        with self._lock:
+            if key not in self._fifos:
+                return 0
+            return len(self._fifos[key]) + self._inprogress_counts[key]
+
+    @property
+    def deferred_size(self) -> int:
+        """Total number of deferred puts not yet processed."""
+        with self._dlock:
+            return len(self._deferrals)
+
+    def _new_fifo(self, key: Hashable) -> deque:
+        """Create and return a new FIFO named `key`.
+
+        A FIFO named `key` must not yet exist!
+
+        Adds `key` to both `_fifos` and `_inprogress_counts`.
+
+        The new FIFO is also added to the zero bin of
+        `_keys_by_inprogress`.
+
+        Parameters
+        ----------
+        key : hashable
+            The name of the FIFO to create.
+
+        Returns
+        -------
+        fifo : deque
+            The newly created FIFO
+        """
+        fifo = deque()
+        self._fifos[key] = fifo
+        self._inprogress_counts[key] = 0
+        self._keys_by_inprogress[0].add(key)
+        return fifo
+
+    def _put(self, item: Any, key: Hashable) -> None:
+        """Put `item` into the FIFO named `key` without locking.
+
+        Never call this function directly; use put() instead.
+
+        NB: The caller must hold the `_lock` when calling this function!
+
+        Parameters
+        ----------
+        item : anything
+            The item to add to the queue
+        key : hashable
+            The name of the FIFO to which `item` is added
+        """
+        # Create the FIFO, if necessary
+        if key not in self._fifos:
+            fifo = self._new_fifo(key)
+        else:
+            fifo = self._fifos[key]
+
+        # push the task onto the FIFO (right-most end)
+        fifo.append(item)
+        self._total_queued += 1
+
+    def put(self, item: Any, key: Hashable, wait: int = 0) -> None:
+        """Put `item` into the queue by pushing it onto the FIFO named `key`.
+
+        If `wait` <= 0, the item is immediately put into the queue.  This is the
+        default behaviour.
+
+        If `wait` > 0, the put is delayed by _at least_ that many seconds before
+        item is actually added to the queue.  In this case, this call does not
+        wait for the put, but returns immediately.
+
+        If the FIFO named `key` doesn't exist, it is created.
+
+        Parameters
+        ----------
+        item : anything
+            The item to add to the queue
+        key : hashable
+            The name of the FIFO to which `item` is added
+        wait : float, optional
+            The number of seconds to wait before adding `item` to the queue.
+        """
+
+        if wait > 0:
+            # Deferred put
+            with self._dlock:
+                # If joining, silently discard this put
```

Comment:
I'm not convinced that this should be entirely silent, there should probably be some sort of warning that this put will never execute.

Reply:
Maybe, though I don't think alpenhornd ever actually calls this method, because if it did, it would be the last thing it does before crashing.  I think I mostly wrote this for completeness sake and/or unit tests.

In alpenhorn, deferred puts can only happen as the result of a `task` that needs to wait for something to happen.  So, even if alpenhorn were to call this message, I don't think it would do it at a time when tasks are still ongoing (because the worker pool is shutdown before the task queue).

Reply:
At minimum, it can't hurt to change the return type of `put` from `None` to `bool` and indicate the failed put that way.